### ISSUE TITLE
CMS: Audit (technical) tables across VA.gov for 508 header considerations

### ISSUE THREAD

**Issue by wesrowe** [department-of-veterans-affairs/va.gov-cms#10793]
## Description
Public Websites + Sitewide Content teams need identify all tables used in CMS & Front-end content. 

Once identified, SWC / PW will review the audit results to understand what tables:
1.  are misusing tables. PW will provide a list of tables, SWC identify which are in this list / are misusing, and will revise this content to stop using tables
2. are formatted correctly but have the wrong accessibility attributes
    * of these which ones are affected by / can be fixed by TableField patch vs.
    * can be fixed by template changes alone

Technical aspects we want to know about each table (combination of FE audit and this CMS audit):
  - FE template used for page -
  - Does FE use [Design System table component](https://design.va.gov/components/table)?
    - If not, what if any reusable component does it leverage (or is it unique to the template)?
  - Presence in drupal - i.e., is table in a paragraph in Drupal or [the version that shows up in [table audit](https://prod.cms.va.gov/admin/content/audit/table-audit) tab]?
  - What products/content types each table lives in (benefit hubs, VAMC, etc)?  Some may belong to Facilities.

## Scope of ticket
This audit should include: 
* Paragraph tables, at any level of nesting
* WYSIWYG tables, if any -- Dave's perspective is that this has never been possible in the CMS with WYSIWYG editors so may be a no-op.

This audit does not include: 
* Images of tables, if any
* FE hard-coded tables (i.e., ~10 tables in react widgets in the vets-website repo)

Specific updates to Drupal Audit View:
- add "VA.gov URL" field (visible in drupal Detail Page)
- tweak query to capture all tables, including ones that are currently missing
  - Example of table missing from current audit: https://www.va.gov/education/survivor-dependent-benefits/
  - May be best to implement as 2 separate views, one for Nodes, one for Paragraphs(?)

## Acceptance Criteria
- [ ] CSV export capability (one or two) listing every table on VAgov, including the existing audit view fields plus VAgov URL
- [ ] Confirm that [page missed](https://www.va.gov/education/survivor-dependent-benefits/) by previous audit view is not present
- [ ] Hand-off meeting with @laflannery and @rlhecht, who will do subsequent content audit to determine how to resolve/remediate issues

### CMS Team
Please check the team(s) that will do this work.

- [ ] `Program`
- [ ] `Platform CMS Team`
- [ ] `Sitewide Crew`
- [x] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public Websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

**Comment by chri5tia** [comment]
https://github.com/department-of-veterans-affairs/va.gov-cms/issues/10379#issuecomment-1241196709

Discovery ticket next steps around this piece:

> 

> Summary
> The patch needs work.
> 
> - [ ] We need UX design to weigh in on placement and help text for the header checkboxes on the input form. See https://github.com/department-of-veterans-affairs/va.gov-cms/issues/10379#issuecomment-1226821574
> - [ ] We need to extend/rework the patch to actually update the database values with what is inputted.
> - [ ] We need to create custom code to address table_field entity definition mismatch, if still applicable after fixing the patch.
> - [ ] We will need an FE engineer to look at how the template will need to be updated to apply the scope property correctly once given the updated data.
> - [ ] We need to audit the table-field audit to find out why there is a large discrepancy between what is reported in the db for table-field rows (1554) vs. what is reported in the audit (396) in admin/content/audit/table-audit
> @wesrowe is cutting tickets for these

> See table audit diffs here.
> [tableauditdiffs.xlsx](https://github.com/department-of-veterans-affairs/va.gov-cms/files/9421303/tableauditdiffs.xlsx)

https://github.com/department-of-veterans-affairs/va.gov-cms/pull/10426#issuecomment-1226763816
> More notes on table audit testing.
> 
> Benefit Detail pages seem okay:
> [tables-usage-audit (2).csv](https://github.com/department-of-veterans-affairs/va.gov-cms/files/9421311/tables-usage-audit.2.csv)

CMS Knowledge base articles seem okay:
[tables-usage-audit (4).csv](https://github.com/department-of-veterans-affairs/va.gov-cms/files/9421318/tables-usage-audit.4.csv)

Resource and Support pages seem okay:
[tables-usage-audit (6).csv](https://github.com/department-of-veterans-affairs/va.gov-cms/files/9421335/tables-usage-audit.6.csv)

VAMC Detail page needs a closer look:
[tables-usage-audit (8).csv](https://github.com/department-of-veterans-affairs/va.gov-cms/files/9421338/tables-usage-audit.8.csv)

admin/content/audit/table-audit?title=&type=health_care_region_detail_page&moderation_state=All&workbench_access_section__section=All
[randstablediff.xlsx](https://github.com/department-of-veterans-affairs/va.gov-cms/files/9421357/randstablediff.xlsx)

These are all the content types that use this entity.

**Comment by jilladams** [comment]
Notes from talking through this with Wes: 
* We suspect that even though we know a lot of CMS content has tables, we may be able to make most of it accessible by modifying the FE templates (vs with the TableField patch) by assuming that the first row = headers, & formatting it that way.
* The Table Audit view in the CMS may or may not contain all the actual tables in the CMS. Swirt was not confident that it did, but we don't know why exactly, so need to take that into account.

As far as what content contains headers: anecdotally in CSV download today, 171 VAMC detail pages, 7 R&S, 1 Centralized content, 144 benefit detail pages.  (That only accounts for what shows up in the Table audit view, so isn't comprehensive but paints a picture.)

**Comment by chri5tia** [comment]
Current table audit view query:

```
SELECT "node_field_data"."langcode" AS "node_field_data_langcode", "node__field_content_block"."delta" AS "node__field_content_block_delta", "node__field_content_block"."langcode" AS "node__field_content_block_langcode", "node__field_content_block"."bundle" AS "node__field_content_block_bundle", "node__field_content_block"."field_content_block_target_id" AS "node__field_content_block_field_content_block_target_id", "node__field_content_block"."field_content_block_target_revision_id" AS "node__field_content_block_field_content_block_target_revisio", "node_field_data"."created" AS "node_field_data_created", "node_field_data"."nid" AS "nid", "paragraphs_item_field_data_node__field_content_block"."id" AS "paragraphs_item_field_data_node__field_content_block_id"
FROM
{node_field_data} "node_field_data"
LEFT JOIN {node__field_content_block} "node__field_content_block" ON node_field_data.nid = node__field_content_block.entity_id AND node__field_content_block.deleted = '0' AND (node__field_content_block.langcode = node_field_data.langcode OR node__field_content_block.bundle IN ( 'page', 'health_care_region_detail_page', 'documentation_page', 'basic_landing_page', 'centralized_content' ))
INNER JOIN {paragraphs_item_field_data} "paragraphs_item_field_data_node__field_content_block" ON node__field_content_block.field_content_block_target_revision_id = paragraphs_item_field_data_node__field_content_block.revision_id
LEFT JOIN {paragraph__field_table} "paragraphs_item_field_data_node__field_content_block__paragraph__field_table" ON paragraphs_item_field_data_node__field_content_block.id = paragraphs_item_field_data_node__field_content_block__paragraph__field_table.entity_id AND (paragraphs_item_field_data_node__field_content_block__paragraph__field_table.deleted = '0' AND paragraphs_item_field_data_node__field_content_block__paragraph__field_table.langcode = paragraphs_item_field_data_node__field_content_block.langcode)
WHERE "paragraphs_item_field_data_node__field_content_block__paragraph__field_table"."field_table_value" IS NOT NULL
ORDER BY "node_field_data_created" DESC
LIMIT 20 OFFSET 0
```

**Comment by jilladams** [comment]
From discussion with Daniel / Christia / Wes: 
**A new possible pool of tables: WYSIWYG**
We haven't searched richtext fields for instances of tables created via WYSIWYG. Tables created that way won't be affected by TableField fixes, won't appear in the Table Audit view, and won't be fixable via front-end templating, as the front-end is probably just displaying them as CMS spits them out (vs. handling them in any way). 
* WYSIWYG'ed tables aren't possible today with the way rich text fields are configured, but maybe they were in the past? 
* Rich text fields without WYSIWYG: may also include tables somehow, if someone gamed the system somehow
* Images of tables could exist. (Barf.) Those are outside the scope of our audit
(@swirtSJW is this a probability or a rabbit trail, in your opinion?)

For separate follow up: the Table Audit view doesn't satisfy this audit bc it can't check FE files or WYSIWYG fields. However: it should be updated to show all instances of Table Paragraphs, for the future. Ownership of that / ticket, TBD.

**Comment by EWashb** [comment]
@edmund-dunn can you look over this issue and familiarize yourself with the ask from Public Websites? Let me know if you have questions/concerns

**Comment by davidmpickett** [comment]
We'll need design representation at the handoff meeting as well. For right now, that will likely be me.

**Comment by wesrowe** [comment]
@davidmpickett, just wondering what the "handoff meeting" would be about. Also noting that in planning this morning the PW team determined we probably don't need to also audit the FE built site; not sure if that's related.

**Comment by davidmpickett** [comment]
@wesrowe The handoff meeting is listed in the acceptance criteria.

**Comment by wesrowe** [comment]
I'm glad I asked – I altered that AC to no longer reference the FE audit that the team deemed unnecessary.

**Comment by jilladams** [comment]
@EWashb @dawnpruitt could I ask you all to update the [Q4 mural](https://app.mural.co/t/vagov6717/m/vagov6717/1664473870398/95d202be38ecc8ad1a03caab7da043d6cc3356ba?sender=u252e7152800e3981889d1910) with timing for this task / when it lands? We have downstream tasks, and we'll need to understand how they're affected by timing here, once it's moving. Thanks!

**Comment by laflannery** [comment]
@jilladams and @dawnpruitt I have spoken to Edmund and he is starting this now and will reach out if he has any questions or issues.  It is a stretch ticket for this sprint and I have updated the mural accordingly to reflect this.

**Comment by dawnpruitt** [comment]
Thank you both!

**Comment by laflannery** [comment]
@edmund-dunn I reviewed this with Randi today and she pointed out a couple examples of pages that were appearing in the audit but not longer had tables.  I looked and she's correct - I also can't figure out what might be happening - our only guess is that maybe the audit is pulling previous revisions?

- Example 1:
  - https://www.va.gov/resources/helpful-va-phone-numbers/
  - https://prod.cms.va.gov/resources/helpful-va-phone-numbers
  - https://prod.cms.va.gov/admin/content/audit/tables?page=29
- Example 2:
  - https://www.va.gov/resources/choosing-a-decision-review-option/
  - https://prod.cms.va.gov/resources/choosing-a-decision-review-option
  - https://prod.cms.va.gov/admin/content/audit/tables?page=42

Are you able to take a look and see if this audit is surfacing older tables that no longer exist on the live site?
